src/wp-includes/kses.php

```diff
@@ -1319,23 +1319,41 @@ function wp_kses_attr_check( &$name, &$value, &$whole, $vless, $element, $allowe
 	$allowed_attr = $allowed_html[ $element_low ];
 
 	if ( ! isset( $allowed_attr[ $name_low ] ) || '' === $allowed_attr[ $name_low ] ) {
+		$name_in_javascript = _wp_kses_transform_custom_data_attribute_name( $name );
+
 		/*
-		 * Allow `data-*` attributes.
+		 * Only allow a subset of custom data attributes containing
+		 * ASCII alphanumerics, underscore, and hyphens. It may be
+		 * safe to allow others, including those with symbols, emoji,
+		 * and other characters, but as a conservative measure these
+		 * are rejected by default.
+		 *
+		 * Code wanting to allow other custom data attributes should
+		 * explicitly pass them in through `$allowed_html`.
+		 */
+		$is_allowable_custom_attribute = (
+			isset( $name_in_javascript ) &&
+			1 === preg_match( '/^[a-z0-9_-]+$/i', $name_in_javascript )
```

Comment:
I learned that `$` actually matches the end of the line in addition to the end of a string. So be the most correct, this regex should actually use `\z`.

```suggestion
			1 === preg_match( '/^[a-z0-9_-]+\z/i', $name_in_javascript )
```

Reply:
So it's possible that `$name_in_javascript` actually ends with `\n` here, but this won't happen if using `\z` instead of `$`.

https://3v4l.org/VO38c

Reply:
In #9953 I have foregone the PCRE match entirely and this is not a problem there.

Another reminder of why I don’t like `preg_` functions for spec-compliance work. There are lots of subtleties in the flags and syntax, some of which even depend on the version of PHP.

https://github.com/WordPress/wordpress-develop/pull/9953/files#diff-36ccb8763042461a0088263081b299d85be1dc53d4354b823d44989901a55437R53-R55